### PR TITLE
CNV-4721 Make CDI work with namespace resource quota

### PR DESCRIPTION
**Pull request by sjhala-ccs**
These modules and assembly address [https://issues.redhat.com/browse/CNV-4721](https://issues.redhat.com/browse/CNV-4721).

**Question**: Should the procedure re: overriding default CPU and memory requests/limits include a step to verify the changes by retrieving the CDIConfig status? If yes, what is the `oc` command to check CDIConfig status and what is the expected response?

Applies to Enterprise-4.5

Code review requested from Adam Litke and Alexander Wels (tagged in Jira)
QE review requested from Ying Cui (tagged in Jira)

### CODE REVIEW

**Comment by demilyc** [comment]
lgtm

File: modules/virt-about-cpu-and-memory-quota-namespace.adoc [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/virtual_disks/virt-configuring-cdi-for-namespace-resourcequota.adoc
+
+[id="virt-about-cpu-and-memory-quota-namespace_{context}"]
+= About CPU and memory quotas in a namespace
+
+A resource quota, defined by the `ResourceQuota` object imposes restrictions on
+a namespace that limit the total amount of compute resources that can be
```
Comment (ousleyp) [inline modules/virt-about-cpu-and-memory-quota-namespace.adoc:9]: I think this could be rephrased to be more customer-focused and minimalistic. 

Suggestion: You can use a _resource quota_ to limit the consumption of compute resources within a namespace. Define a `ResourceQuota` object to set restrictions on [insert specific types of resources here].

File: modules/virt-about-cpu-and-memory-quota-namespace.adoc [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+consumed by resources within that namespace.
+
+The `CDIConfig` object defines the user configuration for CDI. The CPU and
```
Comment (ousleyp) [inline modules/virt-about-cpu-and-memory-quota-namespace.adoc:12]: s/for CDI/for the Containerized Data Importer (CDI).

Since this module could appear outside of this assembly, you need to write out the acronym here, too.

File: modules/virt-about-cpu-and-memory-quota-namespace.adoc [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+memory request/limit values for the `CDIConfig` object are set to a default
```
Comment (ousleyp) [inline modules/virt-about-cpu-and-memory-quota-namespace.adoc:13]: I think the forward slash (request/limit) could be hard to translate.

File: modules/virt-about-cpu-and-memory-quota-namespace.adoc [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+value of 0.
+This ensures that pods created by CDI that make no compute resource requirements
```
Comment (ousleyp) [inline modules/virt-about-cpu-and-memory-quota-namespace.adoc:15]: s/pods/Pods (https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#api-object-formatting)

File: modules/virt-about-cpu-and-memory-quota-namespace.adoc [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+are given the default values and allowed to run in a namespace that is restricted
```
Comment (ousleyp) [inline modules/virt-about-cpu-and-memory-quota-namespace.adoc:16]: s/and allowed/and are allowed

File: modules/virt-editing-cdi-cpu-and-memory-defaults.adoc [hunk 6]
```diff
@@ -0,0 +1,50 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/virtual_disks/virt-configuring-cdi-for-namespace-resourcequota.adoc
+
+[id="virt-editing-cdi-cpu-and-memory-defaults_{context}"]
+= Editing the `CDIConfig` object to override CPU and memory defaults
+
+Modify the default settings for CPU and memory requests and limits for your
+use case by editing the spec attribute of the `CDIConfig` object.
```
Comment (ousleyp) [inline modules/virt-editing-cdi-cpu-and-memory-defaults.adoc:9]: I think "spec" should have backticks here.

File: modules/virt-editing-cdi-cpu-and-memory-defaults.adoc [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+.Procedure
+
+. Use the `oc` client to edit the `cdiconfig/config` and edit the
```
Comment (ousleyp) [inline modules/virt-editing-cdi-cpu-and-memory-defaults.adoc:13]: In this step, you're just doing the first part, so this should stop at "Use the `oc` client to edit the `cdiconfig/config`:" 

Most of our procedure steps have phrasing more like this, though: "Edit the `cdiconfig/config` by running the following command:"

You can also add a `.Prerequisites` section to this module that says:
* Install the OpenShift Command-line Interface (CLI), commonly known as `oc`.

File: modules/virt-editing-cdi-cpu-and-memory-defaults.adoc [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+`spec: podResourceRequirements` to change the default CPU and memory
+requests/limits.
++
+----
+$ oc edit cdiconfig/config
+----
+
+. Update the `spec` property of the `CDIConfig` object
```
Comment (ousleyp) [inline modules/virt-editing-cdi-cpu-and-memory-defaults.adoc:21]: I'd rephrase this to contain the info that was originally in the first step (something like "Change the default CPU and memory limits and requests by editing the `spec.podResourceRequirements` fields:")

File: modules/virt-editing-cdi-cpu-and-memory-defaults.adoc [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+      memory: "1Gi"
+    requests:
+      cpu: "1"
+      memory: "250Mi"
+...
+----
+
+. Save and exit the editor to update the `CDIConfig` object.
+
+. Use the `oc` client to retrieve the `CDIConfig` status and verify your changes.
```
Comment (ousleyp) [inline modules/virt-editing-cdi-cpu-and-memory-defaults.adoc:46]: You can add this under a `.Verification steps` section, and as mentioned earlier, I'd phrase the step more like "View the `CDIConfig` status by running the following command:"

**Review by ousleyp — COMMENTED** [review]
Overall, this is looking good, but I have a few suggestions and style guide notes. Thanks, Shikha!

File: modules/virt-about-cpu-and-memory-quota-namespace.adoc (hunk 1 above)
Comment (sjhala-ccs) [reply modules/virt-about-cpu-and-memory-quota-namespace.adoc:9]: In the context of this module, the compute resource quota is already defined with default values. So, maybe it is out of scope to provide a more detailed explanation on the types of resources that you can configure? 

I could add a .Additional Resources section and link to this assembly: https://docs.openshift.com/container-platform/4.4/applications/quotas/quotas-setting-per-project.html. Let me know what you think.

Comment (ousleyp) [reply modules/virt-about-cpu-and-memory-quota-namespace.adoc:9]: > @ousleyp In the context of this module, the compute resource quota is already defined with default values in place. So, maybe it is out of scope to provide a more detailed explanation on the types of resources that you can configure with the `ResourceQuota` object?

Understood! 

> 
> I could add a `.Additional Resources` section and link to this assembly in the OpenShift docs: https://docs.openshift.com/container-platform/4.4/applications/quotas/quotas-setting-per-project.html.
> 
> Let me know what you think.

This sounds like a good idea to me. I'm glad you found that! Let me know if you have any trouble with the xref.
